### ISSUE TITLE
Duplicate dependency-management clause in powertac-core prevents release processing

### ISSUE THREAD

**Issue by jecollins** [powertac/powertac-server#1013]
Here's the trace from mvn release:perform:
```
    [WARNING] Failed to build parent project for org.apache.geronimo.genesis.config:config:pom:1.1
    [WARNING] Failed to build parent project for org.apache.geronimo.genesis.config:project-config:pom:1.1
    [WARNING] Failed to build parent project for org.apache.geronimo.specs:specs:pom:1.2
    [WARNING] Failed to build parent project for org.apache.geronimo.specs:geronimo-jta_1.0.1B_spec:jar:1.1.1
    [WARNING] Failed to build parent project for org.apache.geronimo.genesis.config:config:pom:1.1
    [WARNING] Failed to build parent project for org.apache.geronimo.genesis.config:project-config:pom:1.1
    [WARNING] Failed to build parent project for org.apache.geronimo.specs:specs:pom:1.2
    [WARNING] Failed to build parent project for org.apache.geronimo.specs:geronimo-jta_1.0.1B_spec:jar:1.1.1
    [WARNING] Unable to create Maven project for org.apache.geronimo.specs:geronimo-jta_1.0.1B_spec:jar:1.1.1 from repository. 
org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
    [WARNING] 'build.pluginManagement.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-one-plugin @ org.apache.geronimo.genesis.config:project-config:1.1, /home/jcollins/.m2/repository/org/apache/geronimo/genesis/config/project-config/1.1/project-config-1.1.pom, line 154, column 17
    [WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-compiler-plugin is missing. @ line 18, column 15
    [ERROR] 'distributionManagement.status' must not be specified. @ line 30, column 13
```

**Comment by jecollins** [comment]
Commenting out the geronimo-jta clause under dependency-management appears to have solved this problem. However, now I'm getting another error:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.7.1:deploy (default-deploy) on project powertac-parent: Wagon protocol 'https' doesn't support directory copying -> [Help 1]
```
Probably this has to do with the sonatype-staging-url, but I'm not sure how to fix it.
```
<sonatype.staging.url>https://oss.sonatype.org/service/local/staging/deploy/maven2/</sonatype.staging.url>
```

**Comment by erikkemperman** [comment]
[here](
https://stackoverflow.com/questions/27602445/maven-site-deploy-issue) folks suggest trying the dav protocol instead of http. But it’s weird, the same config worked before, right?

**Comment by jecollins** [comment]
Yes, I saw that. I have to admit, though, that I did not understand it. It seems that it's more than just adding `dav:` to the URL. Also, I just finished scrolling through all the commits to powertac-core since the last release, and I cannot find a change that would have caused this. So yes, as far as I can see, this worked for the last release in April 2018.

**Comment by erikkemperman** [comment]
Yes they write it needs an additional “extension” but I’m not sure what that means.

Have you tried regular http (no s)?

Is there any extra logging (maybe run maven more verbosely) that might provide clues?

**Comment by jecollins** [comment]
I have not tried using http, and this is an area it's really difficult to "experiment" with, because it's the final stage of release. You have to completely undo the release to try again. Some of that is automated with release:rollback, but that does not seem to clean up the tags, and I'm not sure what else. It takes at least a couple hours per cycle, and of course the commit record is getting filled up with junk. Perhaps I should do a rebase before trying again? I haven't yet done the rollback; I want a clear idea of what to try next before spending the time on that.

**Comment by erikkemperman** [comment]
Yes, git rebase -i <hash> will let you get rid of bogus commits made by the plugin, but not the tags, git branch -D <tag> will delete local tags, git push origin :<tag> will delete tags already pushed to github.

Doesn’t the plugin have a dry-run feature or something like that, might make this trial and error just a little bit less annoying.

**Comment by jecollins** [comment]
Yes, there is a dry run feature, but it would not exercise the staging upload. As far as I can tell, everything else is working correctly.

**Comment by jecollins** [comment]
OK, I renamed the branch called 'failed-release' at the end of the big mess from the failed release attempts, and made a branch called 'master' at the point where it was still good, then cherry-picked the removal of the geronimo-jta library and pushed that as master. Then I created a branch release-1.6 from that, inserted the dav extension and updated the staging url with the dav: prefix.

It fails in release:perform at an early stage: 
```
    Uploading to sonatype-nexus-staging: dav:http://oss.sonatype.org/service/local/staging/deploy/maven2/org/powertac/powertac-parent/1.6.0/powertac-parent-1.6.0.pom
    [INFO] ------------------------------------------------------------------------
    [INFO] Reactor Summary:
    [INFO] 
    [INFO] Power TAC parent ................................... FAILURE [  7.868 s]
    [INFO] Power TAC aspect-oriented programming .............. SKIPPED
    [INFO] Power TAC common domain model ...................... SKIPPED
    [INFO] Power TAC broker-core .............................. SKIPPED
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 8.632 s
    [INFO] Finished at: 2019-03-17T15:42:16-05:00
    [INFO] Final Memory: 28M/301M
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy (default-deploy) on project powertac-parent: Execution default-deploy of goal org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy: java.lang.NoSuchMethodError: org.apache.maven.wagon.shared.http.AbstractHttpClientWagon.getBufferCapacityForTransfer(J)I
    [ERROR] -----------------------------------------------------
    [ERROR] realm =    plugin>org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1
    [ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
    [ERROR] urls[0] = file:/home/jcollins/.m2/repository/org/apache/maven/plugins/maven-deploy-plugin/3.0.0-M1/maven-deploy-plugin-3.0.0-M1.jar
    [ERROR] urls[1] = file:/home/jcollins/.m2/repository/org/sonatype/sisu/sisu-inject-bean/1.4.2/sisu-inject-bean-1.4.2.jar
    [ERROR] urls[2] = file:/home/jcollins/.m2/repository/org/sonatype/sisu/sisu-guice/2.1.7/sisu-guice-2.1.7-noaop.jar
    [ERROR] urls[3] = file:/home/jcollins/.m2/repository/org/sonatype/aether/aether-util/1.7/aether-util-1.7.jar
    [ERROR] urls[4] = file:/home/jcollins/.m2/repository/org/codehaus/plexus/plexus-interpolation/1.14/plexus-interpolation-1.14.jar
    [ERROR] urls[5] = file:/home/jcollins/.m2/repository/org/codehaus/plexus/plexus-component-annotations/1.7.1/plexus-component-annotations-1.7.1.jar
    [ERROR] urls[6] = file:/home/jcollins/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.3/plexus-sec-dispatcher-1.3.jar
    [ERROR] urls[7] = file:/home/jcollins/.m2/repository/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar
    [ERROR] urls[8] = file:/home/jcollins/.m2/repository/org/apache/maven/shared/maven-artifact-transfer/0.10.0/maven-artifact-transfer-0.10.0.jar
    [ERROR] urls[9] = file:/home/jcollins/.m2/repository/org/apache/maven/shared/maven-common-artifact-filters/3.0.1/maven-common-artifact-filters-3.0.1.jar
    [ERROR] urls[10] = file:/home/jcollins/.m2/repository/org/apache/maven/shared/maven-shared-utils/3.1.0/maven-shared-utils-3.1.0.jar
    [ERROR] urls[11] = file:/home/jcollins/.m2/repository/commons-codec/commons-codec/1.11/commons-codec-1.11.jar
    [ERROR] urls[12] = file:/home/jcollins/.m2/repository/commons-io/commons-io/2.5/commons-io-2.5.jar
    [ERROR] urls[13] = file:/home/jcollins/.m2/repository/org/codehaus/plexus/plexus-utils/3.1.0/plexus-utils-3.1.0.jar
    [ERROR] Number of foreign imports: 1
    [ERROR] import: Entry[import  from realm ClassRealm[project>org.powertac:powertac-parent:1.6.0, parent: ClassRealm[maven.api, parent: null]]]
```
I guess I'll clear that out and try again with http:

**Comment by erikkemperman** [comment]
Huh, so that looks like the wagon plugin is being invoked under incorrect
assumptions as to its version. Have we changed maven plugin versions since
last (working) release?

**Comment by jecollins** [comment]
There have been a large number of version changes, from Spring to JUnit. I did not look specifically for maven plugin version changes. The last release was in April 2018. It went smoothly. The clause I added was
```
    <extensions>
        <extension>
            <groupId>org.apache.maven.wagon</groupId>
            <artifactId>wagon-webdav-jackrabbit</artifactId>
            <version>3.3.2</version>
        </extension>
    </extensions>
```
This is the most recent version, but it's possible this version is incompatible with maven-release.version 2.5.3.

**Comment by erikkemperman** [comment]
It looks like there were plugin version changes since previous release:
https://github.com/powertac/powertac-core/commit/a78eeb5a34fbbce15bf5a37e3d9a67118b56d16e

However the maven-release plugin version has't changed in two years, and I don't see an explicit dependency on this wagon plugin. We could try adding that (in powertac-core/pom.xml) at the latest version:
https://mvnrepository.com/artifact/org.codehaus.mojo/wagon-maven-plugin

**Comment by erikkemperman** [comment]
I just noticed that I did update the maven-deploy plugin in the commit I mentioned above:
https://github.com/powertac/powertac-core/commit/a78eeb5a34fbbce15bf5a37e3d9a67118b56d16e#diff-600376dffeb79835ede4a0b285078036R82

We could try going back to the previous version: https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-deploy-plugin/2.8.2

**Comment by jecollins** [comment]
It seems the original problem, before we tried to update the wagon plugin, was in maven-site-plugin. I guess that's pushing the javadocs up to the staging server?

**Comment by erikkemperman** [comment]
I don’t know if it pushes anything, but it generates reports. It was added along with the new CI setup, makes sense timing wise.

So then I guess things might work if you simply edit the broken Geronimo POM?

**Comment by jecollins** [comment]
I am not convinced the geronimo pom is broken -- I've been running without the replacement geronimo-jta dependency since that first try.

**Comment by erikkemperman** [comment]
Hmm, I was pretty specific in my note here:
https://github.com/powertac/powertac-core/blob/8311a1f5ae341e8103281fb854cff285251b98e2/pom.xml#L216-L219

And a coincidence you should stumble over the same artifact. But it's always possible I somehow messed up my ~/.m2 I suppose.

Are we now excluding this dependency altogether?

**Comment by erikkemperman** [comment]
Actually, no, there really is an invalid character (back-quote) in this POM, immediately after the "developers" block:
http://central.maven.org/maven2/org/apache/geronimo/specs/specs/1.1/specs-1.1.pom

**Comment by erikkemperman** [comment]
So I think the version override I had is valid and necessary?

**Comment by jecollins** [comment]
But the original error appears to have been caused by this duplication:
```
[WARNING] 'build.pluginManagement.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-one-plugin @ org.apache.geronimo.genesis.config:project-config:1.1, /home/jcollins/.m2/repository/org/apache/geronimo/genesis/config/project-config/1.1/project-config-1
```
Perhaps that's just a warning and not an error, but removing it (commenting it out) let me get farther. Actually, I cleared out my ~/.m2/repository/org/apache first, but that did not fix it.

If you would be willing to make this release happen yourself, I would be very happy indeed. I think in the meantime I'm going to revert to the earlier versions of all those plugins, where it all worked. It won't be on master, but on the release-1.6 branch.

**Comment by erikkemperman** [comment]
I ceased trusting that Maven builds are deterministic a while ago -- it's as if it is trying to be terribly clever about caching things and remembering that it validated something. Which is clever when it works, terrible when it doesn't. I've seen the same maven command produce different results by simply invoking it repeatedly, for non-obvious reasons. Or to me, anyway.

But I think I remember now more or less why I added that comment at the time, I was trying to get maven-site to work; seems natural that it might overlap with maven-release. I got it working at one point, and at the time it seemed the salient change was this version override. But perhaps mvn got the better of me once again.

I wish I would have been able to be of more help here. I'll be on campus tomorrow morning (CET) and will try to look into this if it's still needed. But note that my availability this week is uncertain due to circumstances out of my control.

**Comment by jecollins** [comment]
That did not work. The older maven surefile plugin uses the older junit, so none of our tests are working.

**Comment by jecollins** [comment]
I updated just the maven-surefire-plugin to 2.22.1. The release:prepare step works again, but now we are back where we started. In release:perform while processing powertac-parent, the maven-site-plugin produces
```
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.5.1:deploy (default-deploy) on project powertac-parent: Wagon protocol 'https' doesn't support directory copying -> [Help 1]
```
I guess this did not happen on the 1.5.0 release because we were not uploading the javadocs? I guess I'll take a closer look at the release history to see if I can form another hypothesis about what change is causing this.

**Comment by jecollins** [comment]
One hypothesis is [the addition of the javadoc plugin](https://github.com/powertac/powertac-core/commit/08829b8ecac5f554f1e4a27b04c61e2d8a712004) in November.

**Comment by jecollins** [comment]
Removing the javadoc plugin had no discernable impact. The next hypothesis is the [removal of two plugins from the release profile](https://github.com/powertac/powertac-core/commit/5a5e91cf383358e176562ae07b6e8fc2574021f0).

**Comment by jecollins** [comment]
Removing the two plugins also had no impact. Nor did changing the staging url to http from https. Nor did removing the trailing / from the staging url. I'm out of ideas for now. I am mystified that this worked last year and no longer works. As far as I know, I've tried it after backing out each even slightly related commit since the last release. Of course, we've migrated to new versions of a large number of libraries and plugins over the last year, but the libraries should not matter, and I've tried it with all the old plugins other than the surefire plugin.

**Comment by erikkemperman** [comment]
I think I have resolved this -- the error message was a bit deceptive, although it pointed me to the right place. It appears that the actual problem was a reference to our CI-generated sites (hosted on github at a https URL) which I added to distributionManagement.

Not sure how that happened, I was just following an example if memory serves. Anyway, it tried to deploy the site via that https URL but of course the github backend is not expecting or allowing that.

So, it looks like 1.6.0 is now released properly. It will take a while for things to show op maven central I believe.
